// components/Notification/Notification.js
import 'pnotify/dist/pnotify.css';

import PNotify from 'pnotify';
import { PropTypes } from 'prop-types';
import React from 'react';

const isReady = () => typeof window !== 'undefined';
// const isReady = () => typeof window !== 'undefined' && typeof window.pnotify !== 'undefined';

let readyCheck;

class Notification extends React.Component {

  static propTypes = {
    type: PropTypes.string,
    title: PropTypes.string,
    text: PropTypes.string,
    animation: PropTypes.string,
    delay: PropTypes.number,
    opacity: PropTypes.number,
    shadow: PropTypes.bool,
    hide: PropTypes.bool,
    nonblock: PropTypes.bool,
    desktop: PropTypes.bool,
  };

  static defaultProps = {
    type: 'notice',
    title: '',
    text: '',
    animation: 'show',
    delay: 2000,
    opacity: 1,
    shadow: true,
    hide: true,
    nonblock: true,
    desktop: true,
  };

  constructor(props) {
    super(props);
    this._renderPNotify = this._renderPNotify.bind(this);
    this.state = {
      ready: isReady(),
      widget: null,
    };

    if (!this.state.ready) {
      readyCheck = setInterval(this._updateReadyState.bind(this), 1000);
    }
  }

  componentDidMount() {
    if (this.state.ready) {
      this._renderPNotify();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const render = this.props;

    if (render === 'explicit' && this.state.ready && !prevState.ready) {
      this._renderPNotify();
    }
  }

  componentWillUnmount() {
    clearInterval(readyCheck);
  }

  _updateReadyState() {
    if (isReady()) {
      this.setState({
        ready: true,
      });

      clearInterval(readyCheck);
    }
  }

  _renderPNotify() {
    const widget = new PNotify(this.props);
    this.setState({ widget });
  }

  render() {
    return (
      <div
        className="pnotify-notification"
      >
      </div>
    );
  }
}

export default Notification;
